live/js.go

```diff
@@ -0,0 +1,198 @@
+// When the action is triggered on the client, phx:hide-start is dispatched to the hidden elements. After the time specified by :time, phx:hide-end is dispatched.
+func (js *JS) Hide(opts *HideOpts) {
+	js.add("hide", opts)
+}
+
+// Show shows elements.
+// When the action is triggered on the client, phx:show-start is dispatched to the shown elements. After the time specified by :time, phx:show-end is dispatched.
+func (js *JS) Show(opts *ShowOpts) {
+	js.add("show", opts)
+}
+
+// Toggle toggles element visibility.
+// When the toggle is complete on the client, a phx:show-start or phx:hide-start, and phx:show-end or phx:hide-end event will be dispatched to the toggled elements.
+func (js *JS) Toggle(opts *ToggleOpts) {
+	js.add("toggle", opts)
+}
+
+type hasArgs interface {
+	args() map[string]any
+}
+
+func (js *JS) add(kind string, a hasArgs) {
+	js.ops = append(js.ops, &op{
+		kind: kind,
+		args: a.args(),
+	})
+}
+
+func (js *JS) MarshalJSON() ([]byte, error) {
+	var o [][]any
+	for _, op := range js.ops {
+		o = append(o, []any{
+			op.kind,
+			op.args,
+		})
+	}
+	return json.Marshal(o)
+}
+
+func (js *JS) String() string {
+	s, _ := js.MarshalJSON()
+	return string(s)
+}
+
+// Transition describes a set of CSS class changes over time.
+type Transition struct {
+	// TransitionClass is the CSS transition class(es) to apply for the duration of the transition.
+	TransitionClass string
+	// StartClass is the CSS class(es) that apply at the start of a transition.
+	StartClass string
+	// EndClass is the CSS class(es) that apply at the end of a transition.
+	EndClass string
+}
+
+type HideOpts struct {
+	// To is the DOM selector of the element to hide, or empty to target the interacted element.
+	To string
+	// Transition to apply, if any.
+	Transition *Transition
+	// Time is the duration in milliseconds of the transition, if present; defaults to DefaultTransitionDuration_ms if 0.
+	Time int32
+}
+
+func (o *HideOpts) args() map[string]any {
+	var to *string
+	time := DefaultTransitionDuration_ms
+	transition := &Transition{}
+	if o != nil {
+		to = &o.To
+		if o.Time != 0 {
+			time = o.Time
+		}
+		if o.Transition != nil {
+			transition = o.Transition
+		}
+	}
+	return map[string]any{
+		"time":       time,
+		"to":         to,
+		"transition": transition.phx(),
+	}
+}
+
+type ShowOpts struct {
+	// To is the DOM selector of the element to show, or empty to target the interacted element.
+	To string
+	// Transition to apply, if any.
+	Transition *Transition
+	// Time is the duration in milliseconds of the transition, if present; defaults to DefaultTransitionDuration_ms if 0.
+	Time int32
+	// Display is the CSS display value to set when showing; defaults to "block".
+	Display string
+}
+
+func (o *ShowOpts) args() map[string]any {
+	var to *string
+	time := DefaultTransitionDuration_ms
+	transition := &Transition{}
+	display := "block"
+	if o != nil {
+		to = &o.To
+		if o.Time != 0 {
+			time = o.Time
+		}
+		if o.Transition != nil {
```

Comment:
I don't think you need these if statements. When nil, You'll overwrite nil with nil.  no harm done.

Reply:
We don't want nil, is the thing—`phx-` attributes expect a 3-tuple of empty js arrays if there's no transition (`[[],[],[]]`)

Reply:
I suppose I could do an equivalent check in `Transition.phx` instead, and return the empty state there if it's nil

Reply:
I'll do that.